File: python/a12.py
# A12 Printer
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38444341

n, k = map(int, input().split())
a = list(map(int, input().split()))

def is_fulfilled(t: int) -> bool:
    return sum(t // aa for aa in a) >= k

def search(y: int) -> int:
    l, r = 1, 10 ** 9
    while l < r:
        m = (l + r) // 2
        if is_fulfilled(m):
            r = m
        else:
            l = m + 1
    return l # == r

print(search(k))

File: rust/tessoku-book/src/bin/a12.rs
// A12 Printer
// submision: https://atcoder.jp/contests/tessoku-book/submissions/38445121

use proconio::input;

fn main() {
    input!{
        n: usize,
        k: u64,
        a:[u64; n],
    }
    let mut l: u64 = 1;
    let mut r: u64 = 10_u64.pow(9);
    while l < r {
        let m = (l + r) / 2;
        let is_fulfilled: bool = {
            let mut sum = 0;
            for aa in &a {
                sum += m / aa;
            }
            sum >= k
        };
        if is_fulfilled {
            r = m;
        } else {
            l = m + 1;
        }
    }
    println!("{}", l); // l == r
}
